### DISCUSSION TITLE
Airflow Deployment to Kubernetes/ Openshift

### DISCUSSION THREAD

**Discussion by mavenzer** [apache/airflow#27315] [Q&A]
Hi, I have been using Airflow for more than a year. We used a custom docker image and faced many issues along the way. We are planning to move towards the official image with some modifications due to security constraints set up by our IT teams. Like changing the python images based on our artifactory and some minor changes.

I know one of the easiest ways to use Helm charts for deploying Airflow, But we can't use it in our organization. 

We plan to migrate to Airflow 2.3 and deploy it in Openshift; what is the best way to deploy it in Kubernetes-based environments?

Any guidance/ advice would be helpful.

Thanks

**Comment by potiuk** [comment]
Despite your reservations, the Official Helm Chart is easiest and the only one I can recommend hearthily. There is no othe solution that matches it in terms of the "reuse" what's there. If you can't use them - you can at least use them as a starting point to build your own way to manage resources, but frankly speaking - the same problems you had when managing custom docker images you will face when managing custom resources on K8S so sooner or later you will likely arrive to the point where you would like to use our chart.

Note that you can extend Airlfow Helm Chart rather easily using it as a subchart of your chart and this way you can add anything your organisation might need. Also Helm Chart (in it's Helm 3+ form as of 2019) is nothing more than a glorified JINJA templating engine and a way how to manage versioning, incremental changes of  resources that are deployed to Kubernetes. Previous versions of Helm required tiler to be installed and some extra resources and complexity. However, as of Helm 3 this all is much simpler. Not sure where your reservations on "Can't use Helm" come from - maybe it is an experience you or your organisatio had from pre-Helm-3 times - so my advice is to challenge this assumption, because you will likely not get anything better and you will save a LOT of effort by using our Charts to start and a LOT of effort for maintining your own resources comparing to using Helm. I cannot think of any reason why an organization would not want to use Helm to manage their applications deployed on K8S. Unless you have other solutions that you want to stick with, I think Helm is as good as it can be and nothing beats it.

Another note. Choosing official image/Dockerfile is a good idea. The official Airlfow Dockerfile (not necessarily image) has indeed great customisation capabilities - I heartily recommend to read the documentation we have on extending original images and building customized ones using our Dockerfile:  https://airflow.apache.org/docs/docker-stack/build.html. Using different base images and other "enterprise" options (such as building image using pre-vetted dependencies, using custom registries and many more) are very well supported in the Dockerfile of ours. And any feedback, comments, improvement proposal are most welcome there as well in case you find some cases that might need additional support. PRs are always welcome.

You can also watch my [talk](https://www.youtube.com/watch?v=wDr3Y7q2XoI) - it was given 2 years ago but it still holds today. Pretty much all of the topics and assumptions I hold then are still valid today and it's a good source of information on the challenges and approach we took when preparing the official image/Dockerfile. Things got better and simpler since (for example you do not need to have whole Airlfow repo in order to build your custom image - the Dockerfile of ours is "standalone" now).

**Comment by mavenzer** [comment]
Thanks @potiuk  I'm increasingly convinced that moving forward, Helm is the easiest way to set up Airflow. Thanks for your guidance.
